vod/subtitle/dfxp_format.c

```diff
@@ -481,192 +504,343 @@ dfxp_append_string(u_char* p, u_char* s)
+			if (vod_strcmp(node->name, DFXP_ELEMENT_SPAN) != 0 ||
+				node->children == NULL ||
+				depth >= DFXP_MAX_STACK_DEPTH)
 			{
 				break;
 			}
+			
+
+			stack[depth].node = node->next;
+			stack[depth].flag = lflag;
 
-			node_stack[node_stack_pos++] = cur_node->next;
-			cur_node = cur_node->children;
+			lflag = dfxp_add_textflags(node, flag);
+
+			p = dfxp_append_tag(p, lflag, stack[depth].flag, 0, appendfunc);  /* open tag */
+
+			depth++;
+			node = node->children;
 			continue;
 
 		default:
 			break;
 		}
 
-		cur_node = cur_node->next;
+		node = node->next;
 	}
 
 	return p;
 }
 
+#define DECORATION_SCRATCH_SPACE (128)
+
 static vod_status_t
-dfxp_get_frame_body(
-	request_context_t* request_context, 
-	xmlNode* cur_node, 
-	vod_str_t* result)
+dfxp_get_frame_body(request_context_t* ctx, xmlNode* node, style *style, vod_str_t* result)
 {
-	size_t alloc_size;
-	u_char* start;
-	u_char* end;
-	
-	// get the buffer length
-	alloc_size = dfxp_get_text_content_len(cur_node);
-	if (alloc_size == 0)
-	{
+	size_t alloc_size = (size_t) dfxp_append_text_content(node, 0, style->flag.decoration, dfxp_fake_append_string);
+	if (alloc_size == 0) {
 		return VOD_NOT_FOUND;
 	}
+	alloc_size += DECORATION_SCRATCH_SPACE;
 
-	alloc_size += 3;		// \n * 3
-
-	// get the text content
-	start = vod_alloc(request_context->pool, alloc_size);
-	if (start == NULL)
-	{
-		vod_log_debug0(VOD_LOG_DEBUG_LEVEL, request_context->log, 0,
-			"dfxp_get_frame_body: vod_alloc failed");
+	u_char* start = vod_alloc(ctx->pool, alloc_size);
+	if (start == NULL){
+		vod_log_debug0(VOD_LOG_DEBUG_LEVEL, ctx->log, 0,"dfxp_get_frame_body: vod_alloc failed");
 		return VOD_ALLOC_FAILED;
 	}
 
-	start++;	// save space for prepending \n
+	u_char* end = dfxp_append_style(start, style);
 
-	end = dfxp_append_text_content(cur_node, start);
-	if ((size_t)(end - start + 2) > alloc_size)
-	{
-		vod_log_error(VOD_LOG_ERR, request_context->log, 0,
-			"dfxp_get_frame_body: result length %uz exceeded allocated length %uz",
-			(size_t)(end - start + 2), alloc_size);
-		return VOD_UNEXPECTED;
+	*end++ = ' ';
+	u_char* textstart = end;
+	end = dfxp_append_text_content(node, end, style->flag.decoration, dfxp_append_string);
+	if (textstart[0] != '\n') {
+		textstart[-1] = '\n';
 	}
+	*end++ = '\n';
+	*end++ = '\n';
 
-	// trim spaces
-	for (;;)
-	{
-		if (start >= end)
-		{
-			return VOD_NOT_FOUND;
-		}
-
-		if (!isspace(start[0]))
-		{
-			break;
-		}
-
-		start++;
-	}
-
-	for (;;)
-	{
-		if (start >= end)
-		{
-			return VOD_NOT_FOUND;
-		}
-
-		if (!isspace(end[-1]))
-		{
-			break;
-		}
-
-		end--;
+	if ((size_t)(end - start) > alloc_size) {
+		vod_log_error(VOD_LOG_ERR, ctx->log, 0,"dfxp_get_frame_body: result length %uz exceeded allocated length %uz",(size_t)(end - start + 2), alloc_size);
+		return VOD_UNEXPECTED;
 	}
 
-	// add leading/trailing newlines
-	start--;
-	*start = '\n';
-	*end++ = '\n';
-	*end++ = '\n';
+	// trim spaces
+//	for (end > start && isspace(end[-1])) {
```

Comment:
Curious if you need to do this space trimming before applying the tag styles?

Reply:
There are a few edge cases between <span>, preservespace, and naked "<p>" tags with text inside. I think we will have to trim leading and trailing whitespace to make the output similar to Viacom.